lib/auth/auth_with_roles_test.go

```diff
@@ -60,6 +64,75 @@ func TestSSOUserCanReissueCert(t *testing.T) {
 	require.NoError(t, err)
 }
 
+// TestGenerateDatabaseCert makes sure users and services with appropriate
+// permissions can generate certificates for self-hosted databases.
+func TestGenerateDatabaseCert(t *testing.T) {
+	ctx := context.Background()
+	srv := newTestTLSServer(t)
+
+	// This user's role doesn't have "create" permission for "db_cert".
+	userWithoutAccess, _, err := CreateUserAndRole(srv.Auth(), "user-without-access", []string{"role1"})
+	require.NoError(t, err)
+
+	// This user's role has "create" permission for "db_cert".
+	userWithAccess, role, err := CreateUserAndRole(srv.Auth(), "user-with-access", []string{"role2"})
+	require.NoError(t, err)
+	role.SetRules(types.Allow, append(role.GetRules(types.Allow), services.NewRule(types.KindDatabaseCert, services.RW())))
+	require.NoError(t, srv.Auth().UpsertRole(ctx, role))
+
+	// Admin user.
+	admin, err := CreateUser(srv.Auth(), "admin", services.NewAdminRole())
+	require.NoError(t, err)
+
+	tests := []struct {
+		desc     string
+		identity TestIdentity
+		err      error
+	}{
+		{
+			desc:     "user can't sign database certs",
+			identity: TestUser(userWithoutAccess.GetName()),
+			err:      trace.AccessDenied(""),
+		},
+		{
+			desc:     "user can sign database certs",
+			identity: TestUser(userWithAccess.GetName()),
+		},
+		{
+			desc:     "user with admin role can sign database certs",
+			identity: TestUser(admin.GetName()),
+		},
+		{
+			desc:     "built-in admin can sign database certs",
+			identity: TestAdmin(),
+		},
+		{
+			desc:     "database service can sign database certs",
+			identity: TestBuiltin(teleport.RoleDatabase),
+		},
+	}
+
+	// Generate CSR once for speed sake.
+	priv, _, err := srv.Auth().GenerateKeyPair("")
+	require.NoError(t, err)
+	csr, err := tlsca.GenerateCertificateRequestPEM(pkix.Name{CommonName: "test"}, priv)
+	require.NoError(t, err)
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			client, err := srv.NewClient(test.identity)
```

Comment:
```suggestion
                        t.Parallel()
			client, err := srv.NewClient(test.identity)
```
Can these subtests be run in parallel?